acceptance-tests/https_ext_crt_list.go

```diff
@@ -0,0 +1,274 @@
+  value: ((ext_crt_list_path))
+- type: replace
+  path: /instance_groups/name=haproxy/jobs/name=haproxy/properties/ha_proxy/ext_crt_list_policy?
+  value: continue
+
+# Generate CA and certificates
+- type: replace
+  path: /variables?/-
+  value:
+    name: common_ca
+    type: certificate
+    options:
+      is_ca: true
+      common_name: bosh
+- type: replace
+  path: /variables?/-
+  value:
+    name: cert_a
+    type: certificate
+    options:
+      ca: common_ca
+      common_name: cert_a.haproxy.internal
+      alternative_names: [cert_a.haproxy.internal]
+- type: replace
+  path: /variables?/-
+  value:
+    name: cert_b
+    type: certificate
+    options:
+      ca: common_ca
+      common_name: cert_b.haproxy.internal
+      alternative_names: [cert_b.haproxy.internal]
+- type: replace
+  path: /variables?/-
+  value:
+    name: cert_c
+    type: certificate
+    options:
+      ca: common_ca
+      common_name: cert_c.haproxy.internal
+      alternative_names: [cert_c.haproxy.internal]
+`
+
+		haproxyBackendPort := 12000
+		extCrtListPath := "/var/vcap/jobs/haproxy/config/ssl/ext-crt-list"
+		haproxyInfo, varsStoreReader := deployHAProxy(haproxyBackendPort, []string{opsfileSSLCertificate}, map[string]interface{}{
+			"ext_crt_list_path": extCrtListPath,
+		})
+
+		dumpHAProxyConfig(haproxyInfo)
+
+		var creds struct {
+			CertA struct {
+				Certificate string `yaml:"certificate"`
+				CA          string `yaml:"ca"`
+				PrivateKey  string `yaml:"private_key"`
+			} `yaml:"cert_a"`
+			CertB struct {
+				Certificate string `yaml:"certificate"`
+				CA          string `yaml:"ca"`
+				PrivateKey  string `yaml:"private_key"`
+			} `yaml:"cert_b"`
+			CertC struct {
+				Certificate string `yaml:"certificate"`
+				CA          string `yaml:"ca"`
+				PrivateKey  string `yaml:"private_key"`
+			} `yaml:"cert_c"`
+		}
+		err := varsStoreReader(&creds)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait for HAProxy to accept TCP connections
+		waitForHAProxyListening(haproxyInfo)
+
+		By("Starting a local http server to act as a backend")
+		closeLocalServer, localPort, err := startLocalHTTPServer(func(w http.ResponseWriter, r *http.Request) {
+			fmt.Fprintln(w, "Hello cloud foundry")
+		})
+		defer closeLocalServer()
+		Expect(err).NotTo(HaveOccurred())
+
+		By(fmt.Sprintf("Creating a reverse SSH tunnel from HAProxy backend (port %d) to local HTTP server (port %d)", haproxyBackendPort, localPort))
+		ctx, cancelFunc := context.WithCancel(context.Background())
+		defer cancelFunc()
+		err = startReverseSSHPortForwarder(haproxyInfo.SSHUser, haproxyInfo.PublicIP, haproxyInfo.SSHPrivateKey, haproxyBackendPort, localPort, ctx)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting a few seconds so that HAProxy can detect the backend server is listening")
+		// HAProxy backend health check interval is 1 second so this should be plenty
+		time.Sleep(5 * time.Second)
```

Comment:
It feels like this entire section is reused in each test. Does it make sense to externalise it as a building block function that can be used by all tests? Basically all tests need:
1. running haproxy
2. some backend
3. ssh tunnel
4. wait until we are good to go

Reply:
Hi @domdom82 I've collapsed the tunnel set up and test server set up into two lines each. We might be able to  abstract them even more and the tests will be shorter but the trade-off is that then someone reading them will have to follow a bunch of function calls to understand how they work rather than being able to read a "list of steps".

Reply:
Yeah. Let's see what happens when we have more tests. I counted 6 occurrences of this section in this PR which I thought may be worthwhile refactoring. If we see 80% of tests have this setup we should probably aim for it to focus less on boiler-plate and more on the actual feature test.